pathology/hovernet/infer_torch.py

```diff
@@ -0,0 +1,256 @@
+        GenerateInstanceBorderd(keys='mask', hover_map_key=HoVerNetBranch.HV, kernel_size=21),
+        GenerateDistanceMapd(keys='mask', border_key='border', smooth_fn=GaussianSmooth()),
+        GenerateWatershedMarkersd(keys='mask', border_key='border', threshold=0.99, radius=3, postprocess_fn=FillHoles(connectivity=2)),
+        Watershedd(keys='dist', mask_key='mask', markers_key='markers')
+    ])
+    if HoVerNetBranch.NC.value in output.keys():
+        type_pred = Activations(softmax=True)(output[HoVerNetBranch.NC.value])
+        type_pred = AsDiscrete(argmax=True)(output[HoVerNetBranch.NC.value])
+
+    pred_inst_dict = post_trans_seg(output)
+    pred_inst = pred_inst_dict['dist']
+
+    inst_id_list = np.unique(pred_inst)[1:]  # exclude background
+
+    inst_info_dict = None
+    if return_centroids:
+        inst_id_list = np.unique(pred_inst)[1:]  # exclude background
+        inst_info_dict = {}
+        for inst_id in inst_id_list:
+            inst_map = pred_inst == inst_id
+            inst_bbox = BoundingRect()(inst_map)
+            inst_map = inst_map[:, inst_bbox[0][0]: inst_bbox[0][1], inst_bbox[0][2]: inst_bbox[0][3]]
+            offset = [inst_bbox[0][2], inst_bbox[0][0]]
+            try:
+                inst_contour = GenerateInstanceContour()(inst_map, offset)
+            except:
+                inst_contour = GenerateInstanceContour()(FillHoles(connectivity=2)(inst_map), offset)
+            inst_centroid = GenerateInstanceCentroid()(inst_map, offset)
+            if inst_contour is not None:
+                inst_info_dict[inst_id] = {  # inst_id should start at 1
+                    "bounding_box": inst_bbox,
+                    "centroid": inst_centroid,
+                    "contour": inst_contour,
+                    "type_probability": None,
+                    "type": None,
+                }
+
+    if output_classes is not None:
+        for inst_id in list(inst_info_dict.keys()):
+            inst_type, type_prob = GenerateInstanceType()(
+                bbox=inst_info_dict[inst_id]["bounding_box"], type_pred=type_pred, seg_pred=pred_inst, instance_id=inst_id)
+            inst_info_dict[inst_id]["type"] = inst_type
+            inst_info_dict[inst_id]["type_probability"] = type_prob
+
+    pred_inst = convert_to_tensor(pred_inst, device=device)
+    pred_type_map = torch.zeros_like(pred_inst)
+    for key, value in inst_info_dict.items():
+        pred_type_map[pred_inst == key] = value['type']
+    pred_type_map = AsDiscrete(to_onehot=5)(pred_type_map)
+
+    if return_binary:
+        pred_inst[pred_inst > 0] = 1
+    return (pred_inst, pred_type_map, inst_info_dict, pred_inst_dict)
+
+
+def main(data_dir, args):
+
+    test_transforms = Compose(
+        [
+            CastToTyped(keys=["image", "label_inst", "label_type"], dtype=torch.float32),
+            AsDiscreted(keys="label_type", to_onehot=5),
+            CenterSpatialCropd(
```

Comment:
you need to deal with WSI or different size input.. looks like hover net input is hardcoded to 270x270